File: C#/More/AIN/DualAINLoop/DualAINLoop.cs
//-----------------------------------------------------------------------------
// DualAINLoop.cs
//
// Demonstrates reading 2 analog inputs (AINs) in a loop from a LabJack.
//
// support@labjack.com
//
// Relevant Documentation:
//
// LJM Library:
//     LJM Library Installer:
//         https://labjack.com/support/software/installers/ljm
//     LJM Users Guide:
//         https://labjack.com/support/software/api/ljm
//     Opening and Closing:
//         https://labjack.com/support/software/api/ljm/function-reference/opening-and-closing
//     Multiple Value Functions (such as eWriteNames and eReadNames):
//         https://labjack.com/support/software/api/ljm/function-reference/multiple-value-functions
//     Timing Functions (such as StartInterval, WaitForNextInterval and
//     CleanInterval):
//         https://labjack.com/support/software/api/ljm/function-reference/timing-functions
//
// T-Series and I/O:
//     Modbus Map:
//         https://labjack.com/support/software/api/modbus/modbus-map
//     Analog Inputs:
//         https://labjack.com/support/datasheets/t-series/ain
//-----------------------------------------------------------------------------

using System;
using LabJack;

namespace DualAINLoop
{
    class DualAINLoop
    {
        static void Main(string[] args)
        {
            DualAINLoop dal = new DualAINLoop();
            dal.performActions();
        }

        public void showErrorMessage(LJM.LJMException e)
        {
            Console.Out.WriteLine("LJMException: " + e.ToString());
            Console.Out.WriteLine(e.StackTrace);
        }

        public void performActions()
        {
            int handle = 0;
            int devType = 0;
            int conType = 0;
            int serNum = 0;
            int ipAddr = 0;
            int port = 0;
            int maxBytesPerMB = 0;
            string ipAddrStr = "";
            int numFrames = 0;
            string[] aNames;
            double[] aValues;
            int errorAddress = -1;
            int intervalHandle = 1;
            int skippedIntervals = 0;

            try
            {
                //Open first found LabJack
                LJM.OpenS("ANY", "ANY", "ANY", ref handle);  // Any device, Any connection, Any identifier
                //LJM.OpenS("T8", "ANY", "ANY", ref handle);  // T8 device, Any connection, Any identifier
                //LJM.OpenS("T7", "ANY", "ANY", ref handle);  // T7 device, Any connection, Any identifier
                //LJM.OpenS("T4", "ANY", "ANY", ref handle);  // T4 device, Any connection, Any identifier
                //LJM.Open(LJM.CONSTANTS.dtANY, LJM.CONSTANTS.ctANY, "ANY", ref handle);  // Any device, Any connection, Any identifier

                LJM.GetHandleInfo(handle, ref devType, ref conType, ref serNum, ref ipAddr, ref port, ref maxBytesPerMB);
                LJM.NumberToIP(ipAddr, ref ipAddrStr);
                Console.WriteLine("Opened a LabJack with Device type: " + devType + ", Connection type: " + conType + ",");
                Console.WriteLine("Serial number: " + serNum + ", IP address: " + ipAddrStr + ", Port: " + port + ",");
                Console.WriteLine("Max bytes per MB: " + maxBytesPerMB);

                //Setup and call eWriteNames to configure AINs.
                if (devType == LJM.CONSTANTS.dtT4)
                {
                    //LabJack T4 configuration
                    //    Resolution index = 0 (default)
                    //    Settling = 0 (auto)
                    aNames = new string[] { "AIN0_RESOLUTION_INDEX", "AIN1_RESOLUTION_INDEX",
                        "AIN0_SETTLING_US", "AIN1_SETTLING_US" };
                    aValues = new double[] { 0, 0, 0, 0};
                    numFrames = aNames.Length;
                    LJM.eWriteNames(handle, numFrames, aNames, aValues, ref errorAddress);
                }
                else
                {
                    //LabJack T7 and T8 configuration

                    // Settling and negative channel do not apply to the T8
                    if (devType == LJM.CONSTANTS.dtT7)
                    {
                        // Negative Channel = 199 (Single-ended)
                        // Settling = 0 (auto)
                        aNames = new string[] {  "AIN0_NEGATIVE_CH","AIN1_NEGATIVE_CH",
                        "AIN0_SETTLING_US", "AIN1_SETTLING_US"};
                        aValues = new double[] { 199, 199, 0, 0 };
                        numFrames = aNames.Length;
                        LJM.eWriteNames(handle, numFrames, aNames, aValues, ref errorAddress);
                    }
                    //    Range = ±10V (T7) or ±11V (T8).
                    //    Resolution index = 0 (default).
                    aNames = new string[] {  "AIN0_RANGE", "AIN1_RANGE",
                        "AIN0_RESOLUTION_INDEX", "AIN1_RESOLUTION_INDEX"};
                    aValues = new double[] { 10, 10, 0, 0 };
                    numFrames = aNames.Length;
                    LJM.eWriteNames(handle, numFrames, aNames, aValues, ref errorAddress);

                }

                //Setup and call eReadNames to read AINs.
                aNames = new string[] {"AIN0", "AIN1"};
                aValues = new double[] {0, 0};
                numFrames = aNames.Length;

                Console.WriteLine("\nStarting read loop.  Press a key to stop.");
                LJM.StartInterval(intervalHandle, 1000000);
                while(!Console.KeyAvailable)
                {
                    LJM.eReadNames(handle, numFrames, aNames, aValues, ref errorAddress);
                    Console.WriteLine("\n" + aNames[0] + " : " + aValues[0].ToString("F4") + " V, " +
                        aNames[1] + " : " + aValues[1].ToString("F4") + " V");
                    LJM.WaitForNextInterval(intervalHandle, ref skippedIntervals);  //Wait 1 second
                    if(skippedIntervals > 0)
                    {
                        Console.WriteLine("SkippedIntervals: " + skippedIntervals);
                    }
                }
            }
            catch (LJM.LJMException e)
            {
                showErrorMessage(e);
            }

            //Close interval and all device handles
            LJM.CleanInterval(intervalHandle);
            LJM.CloseAll();

            Console.WriteLine("\nDone.\nPress the enter key to exit.");
            Console.ReadLine();  // Pause for user
        }
    }
}

File: VB/More/AIN/DualAINLoop/DualAINLoop.vb
'------------------------------------------------------------------------------
' DualAINLoop.vb
'
' Demonstrates reading 2 analog inputs (AINs) in a loop.
'
' support@labjack.com
'------------------------------------------------------------------------------
Option Explicit On

Imports System.Threading
Imports LabJack


Module DualAINLoop

    Sub Main()
        Dim handle As Integer
        Dim devType As Integer
        Dim numFrames As Integer
        Dim aNames() As String
        Dim aValues() As Double
        Dim errAddr As Integer = -1
        Dim intervalHandle As Integer
        Dim skippedIntervals As Integer = 0

        Try
            ' Open first found LabJack
            LJM.OpenS("ANY", "ANY", "ANY", handle)  ' Any device, Any connection, Any identifier
            'LJM.OpenS("T7", "ANY", "ANY", handle)  ' T7 device, Any connection, Any identifier
            'LJM.OpenS("T4", "ANY", "ANY", handle)  ' T4 device, Any connection, Any identifier
            'LJM.Open(LJM.CONSTANTS.dtANY, LJM.CONSTANTS.ctANY, "ANY", handle)  ' Any device, Any connection, Any identifier

            displayHandleInfo(handle)
            devType = getDeviceType(handle)

            ' Setup and call eWriteNames to configure AINs.
            If devType = LJM.CONSTANTS.dtT4 Then
                ' LabJack T4 configuration

                ' AIN0 and AIN1:
                '     Range = +/-10 V. Only AIN0-AIN3 support the +/-10 V range.
                '     Resolution index = 0 (default)
                '     Settling = 0 (auto)
                numFrames = 6
                ReDim aNames(numFrames - 1)
                aNames(0) = "AIN0_RANGE"
                aNames(1) = "AIN0_RESOLUTION_INDEX"
                aNames(2) = "AIN0_SETTLING_US"
                aNames(3) = "AIN1_RANGE"
                aNames(4) = "AIN1_RESOLUTION_INDEX"
                aNames(5) = "AIN1_SETTLING_US"
                ReDim aValues(numFrames - 1)
                aValues(0) = 10
                aValues(1) = 0
                aValues(2) = 0
                aValues(3) = 10
                aValues(4) = 0
                aValues(5) = 0
            Else
                ' LabJack T7 and other devices configuration

                ' AIN0 and AIN1:
                '     Negative Channel = 199 (Single-ended)
                '     Range = +/-10 V
                '     Resolution index = 0 (default)
                '     Settling = 0 (auto)
                numFrames = 8
                ReDim aNames(numFrames - 1)
                aNames(0) = "AIN0_NEGATIVE_CH"
                aNames(1) = "AIN0_RANGE"
                aNames(2) = "AIN0_RESOLUTION_INDEX"
                aNames(3) = "AIN0_SETTLING_US"
                aNames(4) = "AIN1_NEGATIVE_CH"
                aNames(5) = "AIN1_RANGE"
                aNames(6) = "AIN1_RESOLUTION_INDEX"
                aNames(7) = "AIN1_SETTLING_US"
                ReDim aValues(numFrames - 1)
                aValues(0) = 199
                aValues(1) = 10
                aValues(2) = 0
                aValues(3) = 0
                aValues(4) = 199
                aValues(5) = 10
                aValues(6) = 0
                aValues(7) = 0
            End If
            LJM.eWriteNames(handle, numFrames, aNames, aValues, errAddr)

            Console.WriteLine("")
            Console.WriteLine("Set configuration:")
            For i = 0 To numFrames - 1
                Console.WriteLine("  " & aNames(i) & " : " & aValues(i))
            Next

            ' Setup and call eReadNames to read AINs.
            numFrames = 2
            ReDim aNames(numFrames - 1)
            aNames(0) = "AIN0"
            aNames(1) = "AIN1"
            ReDim aValues(numFrames - 1)
            aValues(0) = 0
            aValues(1) = 0

            Console.WriteLine("")
            Console.WriteLine("Starting read loop.  Press a key to stop.")
            intervalHandle = 1
            LJM.StartInterval(intervalHandle, 1000000)
            While Console.KeyAvailable = False
                LJM.eReadNames(handle, numFrames, aNames, aValues, errAddr)
                Console.WriteLine("")
                Console.WriteLine(aNames(0) & " : " & _
                                  aValues(0).ToString("F4") & " V, " & _
                                  aNames(1) & " : " & _
                                  aValues(1).ToString("F4") & " V")
                LJM.WaitForNextInterval(intervalHandle, skippedIntervals)  ' Wait 1 second
                If skippedIntervals > 0 Then
                    Console.WriteLine("SkippedIntervals: " & skippedIntervals)
                End If
            End While
        Catch ljme As LJM.LJMException
            showErrorMessage(ljme)
        End Try

        ' Close interval and all device handles
        LJM.CleanInterval(intervalHandle)
        LJM.CloseAll()

        Console.WriteLine("")
        Console.WriteLine("Done.")
        Console.WriteLine("Press the enter key to exit.")
        Console.ReadLine() ' Pause for user
    End Sub

End Module
